### DISCUSSION TITLE
why readonly field whose type is struct generates ldloca instead of ldflda ?

### DISCUSSION THREAD

**Discussion by leandromoh** [dotnet/runtime#94145] [Q&A] [answered]
A have a struct type `Core`
```c#
struct Core
{
   public Core() => i = 0;
   public int i; 
   public void Inc() { i++;  }
}
```
that is a `readonly` field inside another type `Foo`
```c#
class Foo
{
   public Foo()
   {
     j = 0;   
     core = new Core();
   }
    
   public int j;  
   public int i => core.i;
   public readonly Core core;
     
   public void Inc() 
   { 
       j++; 
       core.Inc(); 
   }
}
```
whoever when I ran the following code the result was different of the expected
```c#
var main = new Foo();
Do();
Do();
Do();

void Do()
{
    main.Inc();
    Console.WriteLine(main.i);
    Console.WriteLine(main.j);
}
```
produces
```c#
0
1
0
2
0
3
```
instead of
```c#
1
1
2
2
3
3
```
I notice that problem is resolved wether 
- change the type of `Core` from `struct` to `class` or 
- maintains it as `struct` but removes the `readonly` from field

analysing the IL code I notice that when it is a `readonly` field (struct) the compiler generates
```IL
ldfld valuetype Core Foo::core 
stloc.0
ldloca.s 0
```
to find the _value_ of a field, load its content and assign it to a local variable it declates by itself, thus making a copy of the field in its initial state (voilà!) and use the copy, so the field value is keeped untouchable each time the method `Foo.Inc` is called.

if I remove the `readonly` of the struct field the compiler just generates
```IL
ldflda valuetype Core Foo::core
```
to find the _address_ of a field and use it, therefore the field has its state changed through the method calls.

ok, thats answer _how_ the program is producing this ouput. the question now is _why_? why the C# compiler makes a defensive copy when the field is `readonly` and field is a `struct`? compiler already guarantees that I can not re-assign the field so I cant see no reason for this.

As final test to check if it could be some IL limitation, I ran IL directly keeping both `initonly` for the field and `ldflda` in the same program and everything cloud works fine.

[complete code (sharplab)](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUHgNwEMwACAWyIEspSBeUqAUwHdSAxAe04AoBKANx4AIr0EixQ3KP5TUAFlIy+eAN55SmitSgA6AJJQAxrI1bUmAJw9KNXVXG4tpC9dt6AVo4C+ePBg5uNTMXAGZAsRD1J2cPelIABgEtEM0jTjAmeOY2AGEMplMY0l9i0hDUcJpgUg9k8uLK0mrm+gA+UnTM+ylzcMyiOE4oABsAT1J8zM6C3ucK8IVSQxM+Bq1VdedND0REZNStLqYDY1ktkrxSvABnYEgjGqmmYMbw5/521oYkheaoGpUA5vFyKFafTZUPb1UreIA) 

[IL program using initonly and ldflda (sharplab)](https://sharplab.io/#v2:DYLglsA0AuIE4FcB2kAmIDUAfAAgNgAIB9AWACgBvcgmggOgGMEBnaAewFsCwlWBDJAwCmBAG5swqAgG0AygE9WQjnQBKyaGA5CAugqUr1STdroBhTgAcIQuLNuiww5uasQ+mtklVDgfAB4eYF7MAILQ0HBgAEYI0EIgIIzscAAUPNAAzABMAJQEALwEqdS0ZQAMAIwE5eU1ABw1dbVNrbWlNLkd9EysnNy80ALCYhJScorxhhpauvpTajOmFhzWwLb2cI7Oi8azK5ZB0RBg0PLhkTFxCUkMKan5RSVkZRXVLVWtAKwALDXZBEqhC+AHYCCCAXhqiC6j8RHgAGYEPAiH5fcH1bplPCZZHomHgv54ACcyKRKIIv2RjQhZPBYJBuKq3S6L1ojBY7C4PH4ghE4kkMnmyl2JjmkxFABEwHwAOZINisJwuSVCWKy2V8aLrC5RWLxRLJNhpUR8YAIIRnSwiCYGUWzPQSlTSuUKpUMFVqhAarU6iJ664AelV6tlPFlAFk2KghMxHsUsbRPh8wcnmummizunRrXAOGBmMxgrxLQQ4EIAI75pCFBNs14EbIiT71Or1PiU3Eg0mMwl45FSJud/s48H47KJmgkvvdxsiL51PDjgfIntd3HTxHIkS9puT4dL5Fd5fTkF/WcL/u0vBSU/r1f7rcUn7Q8/nyGkvAAkH4ono7IMI2dRfF2a77meI6DvO37ohSEGbre/6AdkwGEIeELgRun7kiImRSJkYJDpkdRESRuEkchsLHo+gEQT+4KQuieHblev4Mci1R4MhlGUuU4GwZCgGbriPzRFepJUk+YmHhSzFfgQxH7pkG6EJksGqWCeCqdUmScX8akKYuOkbtUlTosy9avFS5QApUwEbtJPaoVeH7Imhn4mYSj7YSxFllKyZR0AAFnwzBBQQZqysapxBVw5T+LUtStuUfyBoGBCyAAEqElTZqItg1CA5RFUV5AAL7kOQjB+IWBCWFEprxBFcRsBFvBgAQADkAA8UaoAg6wAHydeQVBkGVBBpQQQhIFIbBIgwNXMAQvXRgNQiDZVZDVaFy31WAjUiHwLVtUWBDREICLGpdYC+KgPCnAQAAKcBsLKcB8Bw3RCP48SzcttoLEYYqOnaADy0QAFZCHco3Zr0XIDLyIwCuMwrTHs2ig0DSxCK4qw2HYDhOLG+NrLYADiM22B4QioLqVwGrc9zxs8DY0GmbR8ZZAW0FNEaWkF0bMNm2jQELUj7YdBBBZIaqKGAsoEPwmiAfuqNdb1fA8INAAknV1uzrysFESCytIOgRXAsoi5Z/kEDguLdGNRv84L0bnUIYa8BF0AEKoABqoQ1P4KFcfuU0WDGytgAAXiIdlgqk8V4NEvMNjg6K5XbtA4D8+50MAbAMGay0PX7bNG7Q0jlJbprmpa8jWgQABibCtRw2tIPurL7gAkgAMkQiXFQQwCoEXJd0Mt3NG4Pw+1NkIAECXwDAEjQx8qMgpt2whp3MaDz90PiUgsv4+T3w081MfC/lMSy+r+vPKbyjYyt+3iR94IR85zQ8+JQSAQBUlhb6JQROfCexcr4zzAVUSoj8zTP0GMMEQGQcgfz3iAWUloiBgF/nPE+VQ8DLxwHZdBeQ4GVGiMvEBVCGCQMnnQWe7MAFVHQGPVACJx4DCyACXeiRIZwOyEvB25DjA5HTq8Nh2Qz7ALYKAv+BAZH1EYdA6+LCGwyL4IgteG9UHbykAIkA38GAENYUQ7IED5GKMIQvYiaip6wKUWwnIujkHIzQRI/hn9sG4PwVIsori5FkPKBQwJtBXEMJsXAvCjjmGxKAePbhUgKGYMEXAn4mRSHiL4RE/+RCfiqJiS4wpD9OGXw0ZkmhK8kH6K3hrYxpjzFaKIQuWhCi4FfAQRU9Rzi7G1BAu4+pIw0nGJwdAPBLTpFtOKaE8JXSOF0NKQvL4SSoEMASSswZ1jkk8LGb4oR2zyh4B+DksJ3j8nKKISSDptiLELzwDosslpugTSmjNOaSIxYS2eq9d6n1Eg9QjF3PWI0yDdBwKSZg1oGAymAEgT6Ig4DQBhTDeFiLtAEH3C/AxGsjRwGKPkR22LLIu3Zm7cWHsLre2Wh4f2QcQ4oSEJkCO6Uo4iCLPHAgjRk7+HqFczOPKtoDMSpAvg1stmiqqMM3FDT36AxFMDB06M6AQ2hncfeLM4EnLuTquR5ZoBvMmulT5BB5oEB+R7F6b0PocC1caLa7zTWzXNQtJafzbWfS2jtWqUtabK0rBaPYZpTodWYEIM0dNPZXXLAiW649y6YO+r9T5ANVXKqxqqgOZoLQABUm5CDhpZKaLcE2oFtoFeNd06oIG1E4XhGCjlVvLbW+tgFy5eGAPIMQubG7NwsOWFe10RV83SgLKlFbIXQthRipFZZUWzrNJikQOKUHysFASolDsnZkrZQQCdvyaU8l9gy4O8VmUMH3RymO3KULFHipUH4gr0SYghcc2oeqP09PHhK2UUqHmJVEePTZYAfgAdaXfbJytoApMbT4rBzbANJXFZKzRMy77lKQEIAA7mwKGIz+Tv0HTcAl0yglEPKEA1gcH64WitCIEj6SQAMBHccyo0HDXGo+a6i1VqjG+IJaOmgULlZLoRfOtdnj4MEAmXg7dJLnb7sPdSr2J76WB3PaHOy2dXbsujJyuOCcAR8oYC+4V77pWj1/WhnVP6uG/t7Q3BjBAmPGNY+WHVJDOFwbSSRxIYAdU1K45ZZ101ePfPdgJrBcnAuWdzqSKTr8iOClMQp3dZRyUNkpUetTPsNOMovXZItSjI4GdvSIbIvLH1mf3EKt9OrR7LOlfZv9EGMNAdQ/+9DFGoOQN8945jSHIOJWKSBugYG6C6eQ/fZefBUCoB1c8mj+zBvGOGx12ouzUBtZ65Eohdl+uObo/2xj11mMeZK6Ksysr11v0FP5kxP8rlsMqM85ryHqHLxC2UMLZq+NReY6Y4TJrPW5k0LGbM9UFG2DOIRmTASlNKLoBM+HByYv+OmRVd9f2Isrw9bvH1i1dp1QagGiNFZg2aFDQIM6Eao1SAurGm6d0k0kZTX9CtQonT2izTznNDcC3WmLWUUt5bK3smrTwywdbgANrSXF7oOXhbTrE+i5d86UVorhRrrFSW8Xvy3Q8HdpLMvKcB8e/LftNNMsqNkaI17ytcpEKSPlxJzMNY/U1zp36uvtd651zhoHwN7YKX1mDA2+GueugFnVxSfu0Fx18y1gPHtCfiyJxLSi5X3akGl43im92lfHRbvLdLreFe02pR30dneAkIHyuyHuQcB7FSUlrfvQ/XLvsBnbtmP3Qb2akwbj24vSrG6gYPU2dXlPm4tj9y3YOraj6PnV1iE80CT26lPk7o/li/oIJ1oP/vupJ+z99vq9pk6asddgYbA0M5jddKX90kCPQjE4V6zB5rQFcDGUIiK3aRYLgAAohwBdAtnTAzPqFdrQD9JzumjzpmuKHaNAdcCLuyAjP0DnilmjEgbjNjEqrjGTITJsNsKTAcITFTNhh9PEPTP6IzKRgfGkKzPuJzB8OhlIhyH0NyHdrgQep/oqD/n/kIAAWaArKAeAXTDGPQZcDAQ6iwbWJXA2OwemFmCWiXpOhLpnmrjrhJlilruJiuqbuzDgYYj0CkOliYTQFlq8MrozmXqejbheuUAuDXoZtytVvys3hnptp+pwrtnZrdtJhrIqhjCDKqmgUzGRi9pRt5h9iNrUAaq8qFsfnjvxgQGARATIVEUwSkEfjxsnsTrVB/gwF/sIRymIUAQWHQFkdIVAQwTAUTh6v6jfidLTuGpGusA4czi/kmhmsQZQesETFsCTC4D4AiLIHwAiI3OoOsGEI0egZZPAWmtznaMgYQSoLkRgTQNwYjGYaEQMZjCgTjMcSQcMWQWMecZTNTLQQ0XIdcAocbk8Gwe8BmO0DzPDJyNgXweYaUeUQiL/pUYARIbUVIZAbIQGNEcwc8YbOzKoZmJ8ZZHsT8SEQqkcRETzrkQAKrMByiiGLHQn3AnYuZhG84nEii5F5p/qWhxhKGvFtAAgcGMnfB/A2Q1BSAvjIhCRCRIggggj7g/BSD0RcR9iniLhCTmTARslMmkg/CfjwiNBLjgSuTwSwRsmcFK7pRlp3TaH0Av5tpy4dpv7sBIDdoyYBy2BFheAg72F6miba5zoGGLrq76GrrZ6/H4rMFwlVwyadTgpGzEoZa0C2Gi6aG5a0pOGV4oRfDEjuEVaAjoh8oIg+GNad5BG1J6IHHon4HHGbF0C5FPGxF3zxE+7SpyI2bdZx5+7TaJGzYR7L4YIYmzDXEjHkHjGXRTEzFnBzGxiFkgCWlwDWndwfpAIb4EBb4A674TFdmzHrQLEPFEmOo45pFFEerNnLBuAXHEw7AznTFznzHbHvpAA==)

**Comment by Clockwork-Muse** [comment]
> compiler already guarantees that I can not re-assign the field so I cant see no reason for this.

You are reassigning the field, though.
Keep in mind that structs are value types, and aren't stored on the heap - their data is folded into their containing field in their containing type.  Unlike with a class, where the address of the heap storage wouldn't be reassignable (but the contents could be.

> why the C# compiler makes a defensive copy when the field is a `readonly` struct?

You don't have a `readonly` struct - you'd have to define it on the struct itself (`public readonly struct`).  Keep in mind that the standard recommendation and practice is to use read-only structs _because_ they are value types, and mutating them can cause odd behavior if you're not careful.

**Reply by leandromoh** [reply]
I know structs are value types and stored on the stack, as well mutable structs are not the recommendation, but still can figure out why compiler makes a defensive copy by its own just when I mark the field as `readonly` 😟

> You are reassigning the field, though.

not in my c#, but in the IL generated by the compiler! by my self, I cant make it, even if I want to since the field is marked as `readonly`
in my c# code, the `core` field is assigned exactly once (in the constructor).

![image](https://github.com/dotnet/runtime/assets/11452028/eb6ecb6b-a59e-4499-93e8-b1a5e1d29296)

> You don't have a readonly struct

maybe happened a misunderstood here. i know the type is not a `readonly struct`, I meant that the field is a `struct` and also `readonly`. edited in the post to avoid new confusions.

**Reply by Clockwork-Muse** [reply, marked as answer]
> I know structs are value types and stored on the stack

On the stack isn't important here.  The important bit is that the actual value of the struct is saved into the field itself.  This is why I said "but you are reassigning the field" - the act of mutation would modify the field.
Since you disallowed that (since you added `readonly` to the field, the only way for the compiler to enforce this _with a mutable struct_ is to take the defensive copy.

**Reply by leandromoh** [reply]
oh now I got it !! thank you very much for the explanation @Clockwork-Muse.

**Comment by En3Tho** [comment]
Well, you have to remember that having a field marked read-only means that this piece of memory that this field occupies shouldn't be mutated by conventional means.

For class read-only fields this obviously means that a reference/pointer to the heap shouldn't be mutated/changed.

For struct read-only field this means that whatever fields you have on a struct, they become literally a part of your class body and they can't be mutated.

It works for a class fields because when you call a mutating method you know it will change the class body on the heap but not a read-only reference to it (you can see it as a read-only struct that contains a reference/pointer, actually).

For struct fields compiler has to make a defensive copy because you're calling a mutating method on read-only value itself
